TensorMechanics versions of ASR elasticity, volumetric swelling, and eigenstrain models

bwspenc opened the pull request:
ref #20

agiorla commented:
@bwspenc  Thanks a lot for doing this. I have a few comments, and while some of them are mostly name changes, I am not convinced that this approach for the ConcreteExpansionEigenstrain is the easiest for the user.

I believe a better way would be to have ConcreteASRVolumetricExpansion inherit from ConcreteExpansionEigenstrain. One way to do it would be to add a pure virtual computeQpVolumetricEigenstrain in ConcreteExpansionEigenstrain, call that new method at the start of computeQpEigenstrain. Then ConcreteASRVolumetricExpansion::computeQpVolumetricEigenstrain would simply do the same thing as what ConcreteASRVolumetricExpansion::computeQpProperties is doing right now.

This would remove the need for the user to define a separate ConcreteExpansionEigenstrain in the input file, and also avoid weird scenarios (for example with the current implementation the user could choose a different expansion_type for the ConcreteASRVolumetricExpansion and ConcreteExpansionEigenstrain... Also, it would prevent a situation where the user defines a ConcreteASRVolumetricExpansion but not the associated eigenstrain.

Before going any further in the review, do you think it would be possible to make that change?

bwspenc commented:
@agiorla Good point. I agree with your proposal and will make that change.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/172213/) on df64f41 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/blackbear/docs/PRs/25/site)

This comment will be updated on new commits.

bwspenc commented:
@agiorla I just pushed an updated version that uses your proposed design, which I think is a big improvement over the previous version. This code was all copied straight over from the solid_mechanics implementation, and needed a fair amount of cleanup. I think I have improved it quite a bit (it runs more than 2x faster than the solid_mechanics version now), but it could still use some more critical review.

agiorla commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,278 @@
+
+  params.addParam<Real>("f_compress", -31.0e6, "Compressive strength of concrete (in Pascals)");
+  params.addParam<Real>(
+      "alpha", 1.3333, "Constant for ASR latency time retardation under hydrostatic compression");
```
Maybe this needs a check >= 1 ?

agiorla commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,278 @@
+
+  params.addParam<Real>(
+      "ASR_vol_expansion",
+      0.005,
```
I would like to have no default value here if possible given the quite high variability in concrete reactivity. Idem for tau_c_T0 and tau_L_T0.

agiorla commented on test/tests/concrete_ASR_swelling/asr_confined.i:
```diff
@@ -1,14 +1,17 @@
 # @Requirement F3.50
```
I believe this test to be a little bit too complex (it doesn't provide unit tests of the ASR model features). We can keep it for now, but I would prefer to have elementary tests in the future. I can do it once this has been merged.

bwspenc replied:
I'd like to keep this test for the time being just because I've verified that the new implementation exactly matches the old one. I agree completely that it would be best to have a set of tests targeted at specific aspects of the model, together with more complex tests like this that tie everything together, but let's do that next.

agiorla commented on include/materials/ConcreteASREigenstrain.h:
```diff
@@ -0,0 +1,116 @@
+    Anisotropic
+  };
+  /// Type of expansion
+  const ExpansionType _expansion;
```
I think there is duplication of the same parameter from ConcreteExpansionEigenstrainBase.

bwspenc replied:
You're right -- thanks!

agiorla commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,278 @@
+      "tau_c_T0", 50.0, "Chracteristic ASR time (in days) at reference temprature");
+  params.addParam<Real>("tau_L_T0", 200.0, "Latency ASR time (in days) at reference temprature");
+  params.addParam<Real>("Uc", 5400.0, "Activation energy associated with tau_c");
+  params.addParam<Real>("UL", 9400.0, "Activation energy associated with tau_L");
```
Uc and UL probably need to be strictly > 0

agiorla commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,278 @@
+                        "Set true to turn ASR reaction dependent tensile strength");
+  params.addParam<Real>(
+      "beta_f_tensile", 0.5, "Fractional residual tensile strength at full ASR reaction");
+
```
gamma_tensile, gamma_residual and beta_f_tensile need to be between 0 and 1 (included). beta needs to be >= 0.

bwspenc commented:
I'm not tied to these names at all -- they're just what is in the SolidMechanics model for the most part. I already expanded one or two of them out to make them more descriptive, and I  prefer that in general.  I also think that far too many of the parameters have default values -- I think I would rather have them specified by the user because they are going to be very specific to the concrete mix considered.

bwspenc commented:
@agiorla Sure, if you have opinions, that would be great.  I'm heading out for a little bit right now, but if it works for you, let's talk this afternoon and go through them.

agiorla commented:
@bwspenc I am giving a seminar this afternoon actually. I should be free tomorrow in the afternoon (after 2pm ET here) for a call.

bwspenc commented:
OK, that should work.

bwspenc commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,278 @@
+  Real norm_residual = 10;
+  Real first_norm_residual = 10;
+
+  std::stringstream iter_output;
```
@agiorla I'm definitely open to making this match the procedure used for RIVE. It would change the results, probably by a small amount. I would prefer to keep it the way it is at least for this initial commit just because this exactly matches the solid_mechanics implementation. I've put the parameters related to the iterations in the 'advanced' category, and removed them from the test inputs to avoid widespread use of them.

bwspenc commented:
@agiorla I made all of the name changes we discussed, and added documentation. If you could take another look through this, I'd appreciate it.

agiorla commented on doc/content/documentation/systems/Materials/ConcreteASREigenstrain.md:
```diff
@@ -0,0 +1,265 @@
+                    \end{array}
+        \end{array}\right.
+\end{equation}
+in which $\gamma_t$ is the fraction of the tensile strength beyond which gel is adsorbed by cracks and is a user input parameter in simulation, and
```
I would drop the `Smeared crack` part of the equation as it is currently not implemented.

agiorla commented on doc/content/documentation/systems/Materials/ConcreteASREigenstrain.md:
```diff
@@ -0,0 +1,265 @@
+\label{eqn:asr_strain}
+\Delta\varepsilon^{ASR}=R\{\Delta\varepsilon_{i}^{ASR}\}R^T.
+\end{equation}
+in which $\Delta\varepsilon^{ASR}$ is a 'chemically' imposed incremental strain tensor at each quadrature point, which is used by the solid mechanics model in Grizzly to drive the deformation of concrete structures.
```
This line refers to both the solid mechanics module and Grizzly, which seem like outdated references.

bwspenc commented on doc/content/documentation/systems/Materials/ConcreteASREigenstrain.md:
```diff
@@ -0,0 +1,265 @@
+
+[cite:saouma_constitutive_2006] presented a comprehensive coupled THMC model for ASR swelling based on the model of Ulm et al, and considered the effects of
+stress on both the ASR reaction kinetics and anisotropic volumetric expansion. Saouma and Perotti's ASR model is perhaps the most scientifically rigorous and
+represents the state of the art in modeling ASR expansion, which includes the following features:
```
Sure.  I'll trim this down.

bwspenc commented:
@agiorla Thanks for your review. I addressed your comments, and did some general trimming down of the discussion in the documentation. Maybe it could still use some more trimming. If you could take another look through it to see if I addressed your concerns, I'd appreciate it.

agiorla commented on include/materials/ConcreteExpansionEigenstrainBase.h:
```diff
@@ -0,0 +1,177 @@
+{
+public:
+  ConcreteExpansionEigenstrainBase(const InputParameters & parameters,
+                                   const std::string volumetric_expansion_name);
```
Yes, I am just realizing that it is the best way to make sure the developer doesn't forget it when writing derived classes, so I would say it is wiser to do it like that indeed.

sapitts commented on doc/content/bib/blackbear.bib:
```diff
@@ -1,18 +1,60 @@
 
 @phdthesis{mazars_application_1984,
 	Author = {Mazars, Jacky},
-	Date-Modified = {2018-04-06 23:01:48 +0000},
```
Thanks for pulling these lines out!

sapitts commented on include/materials/ConcreteASREigenstrain.h:
```diff
@@ -0,0 +1,109 @@
+#include "RankTwoTensor.h"
+
+/**
+ * Computes the volumetric expansion eigenstrain due to alkali-silica reaction.
```
Include the reference here too?

sapitts commented on include/materials/ConcreteASREigenstrain.h:
```diff
@@ -0,0 +1,109 @@
+   * Compute the residual for the computation of the ASR reaction extent.
+   * @param qp Current quadrature point id
+   * @param scalar Current iterative value of the ASR reaction extent
+   * @return residual, which is the error in the new ASR reaction extent
```
To a non concrete person, extent isn't as clear as it could be. Do you mean how far along the reaction is?

bwspenc replied:
Yes.  I added an expanded comment to the _ASR_extent variable to explain. It's also well-documented in the md file for this class.

sapitts commented on include/materials/ConcreteASREigenstrain.h:
```diff
@@ -0,0 +1,109 @@
+  /// Coupled value of relative humidity
+  const VariableValue & _rh;
+  /// Power to which relative humidity is raised
+  const Real & _rh_exponent;
```
What's going on with all these abbreviations? Wouldn't `_relative_humidity` be a much nicer variable name?

sapitts commented on include/materials/ConcreteASREigenstrain.h:
```diff
@@ -0,0 +1,109 @@
+  /// Activation energy associated with tau_c
+  Real _Uc;
+  /// Activation energy associated with tau_L
+  Real _UL;
```
I would like to see longer non-abbreviated names for these activation energy variables

bwspenc replied:
These are the names of the variables used in the formulas.  We originally had these short names as the input parameter names, but expanded them out. I think the code is actually more readable if we leave these as U_c, U_L, tau_c_T0, tau_L_T0.

sapitts commented on include/materials/ConcreteASREigenstrain.h:
```diff
@@ -0,0 +1,109 @@
+  /// Whether to include effect of ASR on tensile strength
+  bool _ASR_dependent_tensile_strength;
+  /// Fraction of tensile strength retained at full ASR reaction extent
+  Real _beta_f;
```
`_beta_fraction` please?

bwspenc replied:
f actually doesn't stand for fraction.  Again, the paper this is based on uses 'beta_f' as the variable name, so I'd keep it this way in the code.

sapitts commented on include/materials/ConcreteASRElasticityTensor.h:
```diff
@@ -0,0 +1,51 @@
+  const Real & _poissons_ratio;
+
+  /// Residual fraction of youngs_modulus at full ASR reaction
+  const Real & _beta_e;
```
`_beta_youngs_modulus`?

bwspenc replied:
Same deal.  The paper calls it beta_E.  I'd either use these short names (my preference), or spell them all the way out, which would be pretty long in the code: `_residual_youngs_modulus_fraction`

sapitts commented on include/materials/ConcreteExpansionEigenstrainBase.h:
```diff
@@ -0,0 +1,177 @@
+ *
+ * In the anisotropic case, the following technique is used:
+ * V. Saouma and L. Perotti. Constitutive model for alkali-aggregate
+ * reactions. ACI Materials Journal, 103(3), 2006.
```
Nice description with a reference 👍

sapitts commented on include/materials/ConcreteExpansionEigenstrainBase.h:
```diff
@@ -0,0 +1,177 @@
+  const Real _f_u;
+
+  /// Tensile strength of concrete (Pa)
+  const Real _f_tensile;
```
What does f stand for in all of these variable names?

bwspenc replied:
f_c and f_t are commonly used to denote the compressive and tensile strength of concrete.

sapitts commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,323 @@
+
+  params.addRangeCheckedParam<Real>(
+      "stress_latency_factor",
+      1.3333,
```
Would a fraction be better than a truncated decimal? Or is this value from literature?

bwspenc replied:
It would be better. I'm pretty sure we can't put a mathematical operation here, though.

bwspenc replied:
Turns out you actually can have mathematical operations there.  Cool!

sapitts commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,323 @@
+  params.addParam<Real>(
+      "relative_tolerance", 1e-8, "Relative convergence tolerance for material iteration");
+  params.addParam<Real>(
+      "absolute_tolerance", 1e-20, "Absolute convergence tolerance for material iteration");
```
Didn't these options get taken away in a refactor of the return mapping stress model in tensor mechanics?

bwspenc replied:
This unfortunately doesn't quite fit into the pattern of SingleVariableReturnMappingSolution, so we can't use it right now. As a follow-on to this, I'd like to either make it so I can use that or use a different procedure that doesn't involve iterations.

sapitts commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,323 @@
+                                "Reference temperature for ASR reaction constants.");
+
+  // Note that Fahrenheit is not supported because that would require different parameters for the
+  // times and activation energies
```
Is this note also in the documentation for the class?

bwspenc replied:
The only options are Celsius and Kelvin in the enum, but it looks like the options in the enum don't get dumped out to the webpage.  I'd hate to go spelling out every enum option in the parameter descriptions -- seems like the documentation system should do that for us.

sapitts commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,323 @@
+  else if (temperature_unit == "Kelvin")
+    _temp_offset = 0.0;
+  else
+    mooseError("Unsupported temperature unit");
```
Do you need this check when using Moose Enum?

bwspenc replied:
I think it's a good idea.  It guards against adding an option to the enum and not handling it.

sapitts replied:
I'd recommend a couple of tests for these errors then

sapitts commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,323 @@
+        1.0 - std::exp(_beta) * sig_effective / (1.0 + (std::exp(_beta) - 1.0) * sig_effective);
+    if (gel_absorption_compress > 1.0)
+      gel_absorption_compress = 1.0;
+    if (gel_absorption_compress < 0.0)
```
else if?

bwspenc replied:
done

sapitts commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,323 @@
+    }
+  }
+  else
+    mooseError("Invalid expansion type");
```
Do you need this check if using Moose Enum?

bwspenc replied:
I think it doesn't hurt.

sapitts commented on src/materials/ConcreteASREigenstrain.C:
```diff
@@ -0,0 +1,323 @@
+
+  Real resid = scalar - _ASR_extent_old[qp] -
+               _dt / 86400.0 / tau_c * (1.0 - scalar) / (1.0 + std::exp(-tau_L / tau_c)) *
+                   (scalar + std::exp(-tau_L / tau_c));
```
Wow this is complicated. Would you separate out into a few different variables or add in some parenthesis to make the equation more human readable?

bwspenc replied:
OK, I have an improvement.

sapitts commented on src/materials/ConcreteExpansionEigenstrainBase.C:
```diff
@@ -0,0 +1,257 @@
+  : ComputeEigenstrainBase(parameters),
+    _expansion_type(getParam<MooseEnum>("expansion_type").getEnum<ExpansionType>()),
+    _f_compress(getParam<Real>("compressive_strength")),
+    _f_u(getParam<Real>("expansion_stress_limit")),
```
I think this variable would be a strong candidate for renaming. Maybe something like `expansion_stress_limit`?

bwspenc replied:
I renamed to _sigma_u to match the notation in the paper.

bwspenc commented:
@sapitts I think I addressed your comments.  As far as adding tests of errors goes, which errors did you have in mind?  I don't think it would easy to hit any of the errors with bad user input other than going outside of the ranges on range-checked parameters. They're mostly coding safeguards.

sapitts commented:
I will, however, note that there is no documentation check for blackbear when there probably ought to be one such Civet recipe run.

bwspenc commented:
@sapitts Thanks! I think the doc check is turned on now -- it was just turned off when this PR was first submitted.
